tools/codegen/dest/lazy_ir.py

```diff
@@ -0,0 +1,138 @@
+                scalar_types.append(cpp.argumenttype_type(t, mutable=mutable, binds=binds))
+        elif isinstance(t, OptionalType):
+            if str(t.elem) == 'Tensor':
+                if mutable and not local.use_const_ref_for_mutable_tensors():
+                    types.append(NamedCType(binds, MutRefCType(BaseCType(valueT))))  # TODO: fix this discrepancy
+                    value_types.append(NamedCType(binds, MutRefCType(BaseCType(valueT))))  # TODO: fix this discrepancy
+                else:
+                    types.append(NamedCType(binds, ConstRefCType(OptionalCType(BaseCType(valueT)))))
+                    value_types.append(NamedCType(binds, ConstRefCType(OptionalCType(BaseCType(valueT)))))
+            else:
+                types.append(cpp.argumenttype_type(t, mutable=mutable, binds=binds))
+                scalar_types.append(cpp.argumenttype_type(t, mutable=mutable, binds=binds))
+        elif isinstance(t, ListType):
+            if str(t.elem) == 'Tensor':
+                types.append(NamedCType(binds, BaseCType(valueListT)))
+                value_types.append(NamedCType(binds, BaseCType(valueListT)))
+            elif str(t.elem) == 'Tensor?':
+                types.append(NamedCType(binds, ConstRefCType(ListCType(OptionalCType(BaseCType(valueT))))))
+                value_types.append(NamedCType(binds, ConstRefCType(ListCType(OptionalCType(BaseCType(valueT))))))
+            else:
+                types.append(cpp.argumenttype_type(t, mutable=mutable, binds=binds))
+                scalar_types.append(cpp.argumenttype_type(t, mutable=mutable, binds=binds))
+        else:
+            raise AssertionError(f"unrecognized type {repr(t)}")
+    return types, value_types, scalar_types
+
+@dataclass(frozen=True)
+class LazyIR:
+    backend_index: BackendIndex
+
+    target: Union[
+        Literal[Target.DEFINITION],
+        Literal[Target.DECLARATION],
+    ]
+
+    # TODO(whc) probably use selector instead of building a separate index for ops to codegen
+    # Selector object to determine which operators to generate
+    # registration code for.
+    selector: SelectiveBuilder
+
+    # The namespace that the kernels are written in. This is just `at::native` for in-tree kernels.
+    cpp_namespace: str
+
+    @method_with_native_function
+    def __call__(self, f: Union[NativeFunctionsGroup, NativeFunction]) -> List[str]:
+        func = f.functional.func if isinstance(f, NativeFunctionsGroup) else f.func
+        if func.name in self.backend_index.index:
+            return self.gen(f)
+        else:
+            return []
+
+    def gen(self, f: Union[NativeFunctionsGroup, NativeFunction]) -> List[str]:
+        # for now, we just want one IR class decl and soon after also the method defs
+        # and we use the functional version not out/inplace.
+        func = f.functional.func if isinstance(f, NativeFunctionsGroup) else f.func
```

Comment:
Wouldn't it make more sense to just filter out the inplace/out codegen?

Reply:
(then you would assert that it is functional here)

Reply:
Where would I do the filtering, in the __call__ before it calls gen()?  Or in gen_backend_stubs somehow?  I figured there was a more appropriate way I was missing.

Reply:
you can literally do it here, by returning an empty list. Other examples of filtering:

```
@with_native_function_and_index
def gen_unstructured(f: NativeFunction, backend_index: BackendIndex) -> Optional[str]:
    sig = kernel_signature(f, backend_index)
    metadata = backend_index.get_kernel(f)
    if metadata is None:
        return None
    if "legacy::" in metadata.kernel:
        return None
    else:
        prefix = 'static' if backend_index.external else 'TORCH_API'
        return f"{prefix} {sig.decl(name=metadata.kernel)};"
```